internal/service/alarm_notification_searcher.go

```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2023 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
+compliance with the License. You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under the License is
+distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+implied. See the License for the specific language governing permissions and limitations under the
+License.
+*/
+
+package service
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+
+	"github.com/openshift-kni/oran-o2ims/internal/data"
+	"github.com/openshift-kni/oran-o2ims/internal/jq"
+	"github.com/openshift-kni/oran-o2ims/internal/search"
+)
+
+type subscriptionInfo struct {
+	filters                search.Selector
+	uris                   string
+	consumerSubscriptionId string
+}
+
+// This file contains oran alarm notification serer search for matched subscriptions
+// at 1st step apply linear search
+type alarmSubscriptionSearcher struct {
+	logger *slog.Logger
+	jqTool *jq.Tool
+	//maps with prebuilt selector
+	subscriptionInfoMap *map[string]subscriptionInfo
+	pathIndexMap        *map[string]alarmSubIdSet
+	noFilterSubsSet     *alarmSubIdSet
+
+	//Parser used for the subscription filters
+	selectorParser *search.SelectorParser
+}
+
+func (b *alarmSubscriptionSearcher) SetLogger(
+	value *slog.Logger) *alarmSubscriptionSearcher {
+	b.logger = value
+	return b
+}
+
+func (b *alarmSubscriptionSearcher) SetJqTool(
+	value *jq.Tool) *alarmSubscriptionSearcher {
+	b.jqTool = value
+	return b
+}
+
+func (b *alarmSubscriptionSearcher) SetSelectorParser(
+	parser *search.SelectorParser) *alarmSubscriptionSearcher {
+	b.selectorParser = parser
+	return b
+}
+
+func newAlarmSubscriptionSearcher() *alarmSubscriptionSearcher {
+	return &alarmSubscriptionSearcher{}
+}
+
+func (b *alarmSubscriptionSearcher) build() {
+	b.subscriptionInfoMap = &map[string]subscriptionInfo{}
+	b.pathIndexMap = &map[string]alarmSubIdSet{}
+	b.noFilterSubsSet = &alarmSubIdSet{}
+
+	// Create the filter expression parser:
+	selectorParser, err := search.NewSelectorParser().
+		SetLogger(b.logger).
+		Build()
+	if err != nil {
+		b.logger.Error(
+			"failed to create filter expression parser: ",
+			slog.String("error: ", err.Error()),
+		)
+		return
+	}
```

Comment:
Looks like you are both passing a selector parser as parameter, and then building a new one here. I'd say it is better to build it here, so consider removing the `SetSelectorParser` method from the builder.

Reply:
I am not sure we passed in the selector to handler. The parser was created in the Adapter codes that is part of server package and not accessible to handler in service package. Good thing the parser self just contains logger object and we only duplicate a pointer variable that point to the same logger object. As we are changing not using Adaptor for this handler, I think we need to keep like this now. Please let me know if you have other thought.